platform/jewel/markdown/core/src/main/kotlin/org/jetbrains/jewel/markdown/processing/MarkdownProcessor.kt

```diff
@@ -160,38 +160,52 @@ public class MarkdownProcessor(
             }
         }
 
-        val newBlocks = previousBlocks.subList(0, firstBlock) + updatedBlocks + suffixBlocks
+        val prefixBlocks = if (firstBlock == -1) emptyList() else previousBlocks.subList(0, firstBlock.coerceAtLeast(0))
+        val newBlocks = prefixBlocks + updatedBlocks + suffixBlocks
         currentState = State(rawMarkdown, newBlocks)
 
         return newBlocks
     }
 
+    /**
+     * Identifies the range of `Block` objects in `previousBlocks` affected by changes between `previousText` and
+     * `updatedText`.
+     *
+     * This function is designed to intelligently handle changes occurring near block boundaries. It includes a block
+     * *before* the actual textual change to accommodate potential block merging.
+     */
     @VisibleForTesting
     internal fun findChangedBlocks(
         previousText: String,
-        updatedText: String,
+        fullUpdatedText: String,
         previousBlocks: List<Block>,
     ): FindChangedBlocksResult {
-        val nCharsDelta = updatedText.length - previousText.length
-        val commonPrefix = previousText.commonPrefixWith(updatedText)
+        val previousStartIndexes = previousBlocks.map { block -> block.sourceSpans.first().inputIndex }
+        val previousEndIndexes =
+            previousBlocks.map { block -> block.sourceSpans.last().let { it.inputIndex + it.length } }
```

Comment:
@AlexVanGogen please pay attention to this line: I assume I don't need to take the bottom right node of the tree? (maybe a test with sublists would be good here)

Reply:
More tests is always good :)

Reply:
added as a separate commit